x/staking/keeper/liquid_stake.go

```diff
@@ -0,0 +1,155 @@
+// This is determined by checking if the account is a 32-length module account
+func (k Keeper) AccountIsLiquidStakingProvider(ctx sdk.Context, address sdk.AccAddress) bool {
+	account := k.authKeeper.GetAccount(ctx, address)
+	_, isModuleAccount := account.(*authtypes.ModuleAccount)
+	return isModuleAccount && len(address) == 32
+}
+
+// ExceedsGlobalLiquidStakingCap checks if a liquid delegation would cause the
+// global liquid staking cap to be exceeded
+// A liquid delegation is defined as either tokenized shares, or a delegation from an ICA Account
+// The total stake is determined by the balance of the bonded pool
+// Returns true if the cap is exceeded
+func (k Keeper) CheckExceedsGlobalLiquidStakingCap(ctx sdk.Context, tokens sdk.Int) bool {
+	liquidStakingCap := k.GlobalLiquidStakingCap(ctx)
+	liquidStakedAmount := k.GetTotalLiquidStakedTokens(ctx)
+
+	// Determine the total stake from the balance of the bonded pools
+	bondedPoolAddress := k.authKeeper.GetModuleAddress(types.BondedPoolName)
+	totalStakedAmount := k.bankKeeper.GetBalance(ctx, bondedPoolAddress, k.BondDenom(ctx)).Amount
+
+	// Calculate the percentage of stake that is liquid
+	updatedTotalStaked := sdk.NewDecFromInt(totalStakedAmount.Add(tokens))
+	updatedLiquidStaked := sdk.NewDecFromInt(liquidStakedAmount.Add(tokens))
+	liquidStakePercent := updatedLiquidStaked.Quo(updatedTotalStaked)
+
+	return liquidStakePercent.GT(liquidStakingCap)
+}
+
+// ExceedsValidatorBondCap checks if a liquid delegation to a validator would cause
+// the liquid shares to exceed the validator bond factor
+// A liquid delegation is defined as either tokenized shares, or a delegation from an ICA Account
+// Returns true if the cap is exceeded
+func (k Keeper) CheckExceedsValidatorBondCap(ctx sdk.Context, validator types.Validator, shares sdk.Dec) bool {
+	validatorBondFactor := k.ValidatorBondFactor(ctx)
+	maxValLiquidShares := validator.TotalValidatorBondShares.Mul(validatorBondFactor)
+	return validator.TotalLiquidShares.Add(shares).GT(maxValLiquidShares)
+}
+
+// SafelyIncreaseTotalLiquidStakedTokens increments the total liquid staked tokens
+// if the global cap is enabled and is not surpassed by this delegation
+func (k Keeper) SafelyIncreaseTotalLiquidStakedTokens(ctx sdk.Context, amount sdk.Int) error {
+	// If the cap is disabled, do nothing
+	if !k.GlobalLiquidStakingCapEnabled(ctx) {
+		return nil
+	}
+
+	// Confirm the cap will not be exceeded
+	if k.CheckExceedsGlobalLiquidStakingCap(ctx, amount) {
+		return types.ErrGlobalLiquidStakingCapExceeded
+	}
+
+	// Increment the global total liquid staked
+	k.SetTotalLiquidStakedTokens(ctx, k.GetTotalLiquidStakedTokens(ctx).Add(amount))
+
+	return nil
+}
+
+// DecreaseTotalLiquidStakedTokens decrements the total liquid staked tokens
+// if the global cap is enabled
+func (k Keeper) DecreaseTotalLiquidStakedTokens(ctx sdk.Context, amount sdk.Int) {
+	if k.GlobalLiquidStakingCapEnabled(ctx) {
```

Comment:
How does it work if global cap disabled after some time of enable and looking to enable again? I think state might be broken for this check

Reply:
Other option is to just always count it, but my concern was that if it's disabled, we'd be wasting operations

Reply:
I see - upgrade process would be pretty heavy if it's done on upgrade handler. (Iterating all the delegator accounts)